# HW3B - Pandas Fundamentals

See Canvas for details on how to complete and submit this assignment.

## Introduction

This assignment transitions you from NumPy's numerical array operations to Pandas' powerful tabular data manipulation. While NumPy excels at homogeneous numerical arrays, Pandas is designed for the heterogeneous, labeled data that characterizes most real-world datasets—mixing dates, categories, numbers, and text within the same table.

You'll work with real bike share data from Chicago's Divvy system to answer questions about urban transportation patterns. Through three progressively complex problems—exploring usage patterns, analyzing rider behavior, and conducting temporal analysis—you'll discover why Pandas has become the standard tool for data analysis in Python.

The assignment emphasizes Pandas' design philosophy: named column access, explicit indexing methods (loc/iloc), handling missing data, and method chaining for readable data pipelines. You'll also see how Pandas builds on NumPy while adding the structure and convenience needed for practical data science work.

This assignment should take 3-5 hours to complete.

Before submitting, ensure your notebook:

- Runs completely with "Kernel → Restart & Run All"
- Includes thoughtful responses to all interpretation questions
- Uses clear variable names and follows good coding practices
- Shows your work (don't just print final answers)

### Learning Objectives

By completing this assignment, you will be able to:

1. **Construct and manipulate Pandas data structures**
   - Create DataFrames from dictionaries and CSV files
   - Distinguish between Series and DataFrame objects
   - Set and reset index structures appropriately
   - Understand when operations return views vs copies
2. **Apply explicit indexing paradigms**
   - Use `loc[]` for label-based data access
   - Use `iloc[]` for position-based data access
   - Access columns using bracket notation
   - Explain when each indexing method is appropriate
3. **Diagnose and explore datasets systematically**
   - Use `info()`, `describe()`, `head()`, and `dtypes` to understand data structure
   - Identify missing values with `isna()` and `notna()`
   - Calculate summary statistics across different axes
   - Interpret value distributions with `value_counts()`
4. **Filter data with boolean indexing and queries**
   - Combine multiple conditions with `&`, `|`, and `~` operators
   - Use `isin()` for membership testing
   - Apply `query()` for readable complex filters
   - Understand how index alignment affects operations
5. **Work with datetime data**
   - Parse dates during CSV loading
   - Extract temporal components with the `.dt` accessor
   - Filter data by date ranges
   - Create time-based derived features
6. **Connect Pandas patterns to data analysis workflows**
   - Formulate questions that data can answer
   - Choose appropriate methods for different analysis tasks
   - Interpret results in domain context
   - Recognize when vectorized operations outperform apply()

### Generative AI Allowance

You may use GenAI tools for brainstorming, explanations, and code sketches if you disclose it, understand it, and validate it. Your submission must represent your own work and you are solely responsible for its correctness.

### Scoring

Total of 90 points available, will be graded out of 80. Scores of >100% are allowed.

Distribution:

- Tasks: 48 pts
- Interpretation: 32 pts
- Reflection: 10 pts

Points by Problem:

- Problem 1: 3 tasks, 10 pts
- Problem 2: 4 tasks, 14 pts
- Problem 3: 4 tasks, 14 pts
- Problem 4: 3 tasks, 10 pts

Interpretation Questions:

- Problem 1: 3 questions, 8 pts
- Problem 2: 4 questions, 8 pts
- Problem 3: 3 questions, 8 pts
- Problem 4: 3 questions, 8 pts

Graduate differentiation: poor follow-up responses will result in up to a 5pt deduction for that problem.

## Dataset: Chicago Divvy Bike Share

The dataset you will analyze is based on real trip information from Divvy, Chicago's bike share system. It contains individual trips with start/end times, station information, and rider type.

Dataset homepage: https://divvybikes.com/system-data

Each trip includes:

- Trip start and end times (datetime)
- Start and end station names and IDs
- Rider type (member vs casual)
- Bike type (classic, electric, or docked)

Chicago's Department of Transportation uses this data to optimize station placement, understand usage patterns, and improve service. You'll explore similar questions that real transportation analysts investigate.

## Problems

### Problem 1: Creating DataFrames from Scratch

Before loading data from files, you need to understand how Pandas structures are built. In this problem, you'll create Series and DataFrames manually using Python's built-in data structures. This is a quick warmup to establish the fundamentals.

#### Task 1a: Create a Series

Create a Series called `temperatures` representing daily high temperatures for a week:

- Monday: 72°F
- Tuesday: 75°F  
- Wednesday: 68°F
- Thursday: 71°F
- Friday: 73°F

Use the day names as the index. Print the Series and its data type.

##### Your Code

In [1]:
import pandas as pd

# Create a Series of daily high temperatures
temperatures = pd.Series(
    data=[72, 75, 68, 71, 73],
    index=["Monday", "Tuesday", "Wednesday", "Thursday", "Friday"]
)

# Display the Series and its data type
print(temperatures)
print("\nData type:", temperatures.dtype)


Monday       72
Tuesday      75
Wednesday    68
Thursday     71
Friday       73
dtype: int64

Data type: int64


#### Task 1b: Create a DataFrame from a Dictionary

Create a DataFrame called `products` with the following data:

| product | price | quantity |
|---------|-------|----------|
| Widget  | 19.99 | 100 |
| Gadget  | 24.99 | 75 |
| Doohickey | 12.49 | 150 |

Use a dictionary where keys are column names and values are lists. Print the DataFrame and report its shape.

##### Your Code

In [2]:
import pandas as pd

# Create the DataFrame from a dictionary
products = pd.DataFrame({
    "product": ["Widget", "Gadget", "Doohickey"],
    "price": [19.99, 24.99, 12.49],
    "quantity": [100, 75, 150]
})

# Display the DataFrame and its shape
print(products)
print("\nShape:", products.shape)



     product  price  quantity
0     Widget  19.99       100
1     Gadget  24.99        75
2  Doohickey  12.49       150

Shape: (3, 3)


#### Task 1c: Access DataFrame Elements

Using the `products` DataFrame from Task 1b, extract and print:

1. The `price` column as a Series
2. The `product` and `quantity` columns as a DataFrame (using a list of column names)

##### Your Code

In [3]:
# Access the "price" column as a Series
price_series = products["price"]
print("Price column as Series:")
print(price_series)
print(type(price_series))  # to verify 

print("\n--------------------------\n")

# Access "product" and "quantity" columns as a new DataFrame
subset_df = products[["product", "quantity"]]
print("Product and Quantity columns as DataFrame:")
print(subset_df)
print(type(subset_df))  # to verify



Price column as Series:
0    19.99
1    24.99
2    12.49
Name: price, dtype: float64
<class 'pandas.core.series.Series'>

--------------------------

Product and Quantity columns as DataFrame:
     product  quantity
0     Widget       100
1     Gadget        75
2  Doohickey       150
<class 'pandas.core.frame.DataFrame'>


#### Interpretation

Answer the following questions (briefly / concisely) in the markdown cell below:

1. Data structure mapping: When you create a DataFrame from a dictionary (like in Task 1b), what do the dictionary keys become? What do the values become?
2. Bracket notation: Why does `df['price']` return a Series, but `df[['price']]` return a DataFrame? What's the difference in what you're asking for?
3. Index purpose: In Task 1a, you used day names as the index instead of default numbers (0, 1, 2...). When would a custom index like this be more useful than the default numeric index?

##### Your Answers

### Interpretation

**Data structure mapping:**  
When someone makes the DataFrame from a dictionary, then the dictionary keys implicitly act as column names and the dictionary values implicitly act as data, aligned by position within.

**Bracket notation:**  
`df['price']` returns a Series since it asks for one column.
`df[['price']]` returns a DataFrame for this case. The list is read as a request to "return a table (even if it contains one column)".


**Index purpose:**  
Custom indexes such as day names can be useful when the index values convey metadata, e.g. time periods, categories, identifiers and so on. A custom index provides a semantic way to reference or align data.

### Problem 2: Loading and Initial Exploration

Before starting this problem, make sure you are working in a copy of this file in the `my_repo` folder you created in HW2a. You must also have a copy of the file `202410-divvy-tripdata-100k.csv` in a subdirectory called `data`. That file structure is illustrated below.

```text
~/insy6500/my_repo
└── homework
    ├── data
    │   └── 202410-divvy-tripdata-100k.csv
    └── hw3b.ipynb
```

#### Task 2a: Load and Understand Raw Data

Start by loading the data "as-is" to get a general understanding of the overall structure and how Pandas interprets it by default.

Note on file paths: The provided code uses `Path` from Python's `pathlib` module to handle file paths. Path objects work consistently across operating systems (Windows uses backslashes `\`, Mac/Linux use forward slashes `/`), automatically using the correct separator for your system. The provided code defines `csv_path` which should be used as the filename in your `pd.read_csv` to load the data file.

1. Use `pd.read_csv` to load `csv_path` (provided below) without specifying any other arguments. Assign it to the variable `df_raw`.
2. Use the methods we described in class to explore the shape, structure, types, etc. of the data. In particular, consider which columns represent dates or categories.
3. Note the amount of memory used by the dataset. See the section on memory diagnostics in notebook 07a for appropriate code snippets using `memory_usage`.

##### Your Code

In [12]:
import pandas as pd
import numpy as np
from pathlib import Path

# create a OS-independent pointer to the csv file created by Setup
csv_path = Path('../data/202410-divvy-tripdata-100k.csv')

# load and explore the data below (create additional code / markdown cells as necessary)

# Step 1: load raw data "as-is"
df_raw = pd.read_csv(csv_path)
df_raw.head() 

# Step 2: basic structure & types
print("Shape:", df_raw.shape)
print("\nColumns:", df_raw.columns.tolist())

print("\nDtypes:\n", df_raw.dtypes)

print("\nInfo():")
df_raw.info()

# Step 3: quick descriptive stats (numeric only) 
df_raw.describe()

# Step 4: missingness overview 
na_counts = df_raw.isna().sum().sort_values(ascending=False)
na_counts[na_counts > 0]

# Step 5: memory diagnostics (per 07a approach) 
mem_per_col_mb = df_raw.memory_usage(deep=True) / (1024**2)
total_mem_mb = mem_per_col_mb.sum()
print("Memory usage per column (MB):\n", mem_per_col_mb.sort_values(ascending=False))
print("\nTotal memory usage (MB):", round(total_mem_mb, 2))







Shape: (100000, 13)

Columns: ['ride_id', 'rideable_type', 'started_at', 'ended_at', 'start_station_name', 'start_station_id', 'end_station_name', 'end_station_id', 'start_lat', 'start_lng', 'end_lat', 'end_lng', 'member_casual']

Dtypes:
 ride_id                object
rideable_type          object
started_at             object
ended_at               object
start_station_name     object
start_station_id       object
end_station_name       object
end_station_id         object
start_lat             float64
start_lng             float64
end_lat               float64
end_lng               float64
member_casual          object
dtype: object

Info():
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 13 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   ride_id             100000 non-null  object 
 1   rideable_type       100000 non-null  object 
 2   started_at          100000 non-n

#### Task 2b: Reload with Proper Data Types

1. Repeat step 2a.1 to reload the data. Use the `dtype` and `parse_dates` arguments to properly assign categorical and date types. Assign the result to the variable name `rides`.
2. After loading, use `rides.info()` to confirm the type changes.
3. Use `memory_usage` to compare the resulting size with that from step 2a.3.

##### Your Code

In [15]:
# Dates: parse started_at, ended_at
# Categoricals: minimal obvious categories from Divvy schema
rides = pd.read_csv(
    csv_path,
    parse_dates=["started_at", "ended_at"],
    dtype={
        "rideable_type": "category",
        "member_casual": "category",
    }
)

# Confirm type changes
rides.info()

# Compare memory usage with step 2a.3 (df_raw from earlier)
mem_raw_mb = df_raw.memory_usage(deep=True).sum() / (1024**2)
mem_rides_mb = rides.memory_usage(deep=True).sum() / (1024**2)
print(f"Raw (df_raw) total memory MB:   {mem_raw_mb:.2f}")
print(f"Typed (rides) total memory MB: {mem_rides_mb:.2f}")



<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 13 columns):
 #   Column              Non-Null Count   Dtype         
---  ------              --------------   -----         
 0   ride_id             100000 non-null  object        
 1   rideable_type       100000 non-null  category      
 2   started_at          100000 non-null  datetime64[ns]
 3   ended_at            100000 non-null  datetime64[ns]
 4   start_station_name  89623 non-null   object        
 5   start_station_id    89623 non-null   object        
 6   end_station_name    89485 non-null   object        
 7   end_station_id      89485 non-null   object        
 8   start_lat           100000 non-null  float64       
 9   start_lng           100000 non-null  float64       
 10  end_lat             99913 non-null   float64       
 11  end_lng             99913 non-null   float64       
 12  member_casual       100000 non-null  category      
dtypes: category(2), datetime64[ns]

#### Task 2c: Explore Structure and Missing Data

Using the `rides` DataFrame from Task 2b:

1. Determine the range of starting dates in the dataframe using the `min` and `max` methods.
2. Count the number of missing values in each column. See the section of the same name in lecture 06b.
3. Convert the Series from step 2 to a DataFrame using `.to_frame(name='count')`, then add a column called 'percentage' that calculates the percentage of missing values for each column.

##### Your Code

In [16]:


# Range of starting dates
print("Date range of rides:")
print("Min:", rides["started_at"].min())
print("Max:", rides["started_at"].max())

# Count missing values per column
na_counts = rides.isna().sum()

# Convert to DataFrame and add percentage column
na_df = na_counts.to_frame(name="count")
na_df["percentage"] = (na_df["count"] / len(rides)) * 100

# Display results
na_df


Date range of rides:
Min: 2024-09-30 23:12:01.622000
Max: 2024-10-31 23:54:02.851000


count  percentage
ride_id                 0       0.000
rideable_type           0       0.000
started_at              0       0.000
ended_at                0       0.000
start_station_name  10377      10.377
start_station_id    10377      10.377
end_station_name    10515      10.515
end_station_id      10515      10.515
start_lat               0       0.000
start_lng               0       0.000
end_lat                87       0.087
end_lng                87       0.087
member_casual           0       0.000

#### Task 2d: Create Trip Duration Column and Set Index

Before setting the index, create a derived column for trip duration:

1. Calculate trip_duration_min by subtracting `started_at` from `ended_at`, then converting to minutes using `.dt.total_seconds() / 60`
3. Display basic statistics (min, max, mean) for the new column using `.describe()`
4. Show the first few rows with `started_at`, `ended_at`, and `trip_duration_min` to verify the calculation
5. Set `started_at` as the DataFrame's index. Verify the change by printing the index and displaying the first few rows.

##### Your Code

In [19]:
# Calculate trip duration in minutes
rides["trip_duration_min"] = (rides["ended_at"] - rides["started_at"]).dt.total_seconds() / 60

# Display basic statistics
rides["trip_duration_min"].describe()

# Verify calculation
rides[["started_at", "ended_at", "trip_duration_min"]].head()

# Set started_at as index
rides = rides.set_index("started_at")

# Verify index
print("Index name:", rides.index.name)
print("Index type:", type(rides.index))
rides.head()




Index name: started_at
Index type: <class 'pandas.core.indexes.datetimes.DatetimeIndex'>


ride_id  rideable_type  \
started_at                                                 
2024-09-30 23:12:01.622  67BB74BD7667BAB7  electric_bike   
2024-09-30 23:19:25.409  5AF1AC3BA86ED58C  electric_bike   
2024-09-30 23:32:24.672  7961DD2FC1280CDC   classic_bike   
2024-09-30 23:42:11.207  2E16892DEEF4CC19   classic_bike   
2024-09-30 23:49:25.380  AAF0220F819BEE01  electric_bike   

                                       ended_at         start_station_name  \
started_at                                                                   
2024-09-30 23:12:01.622 2024-10-01 00:20:00.674     Oakley Ave & Touhy Ave   
2024-09-30 23:19:25.409 2024-10-01 00:42:09.933                        NaN   
2024-09-30 23:32:24.672 2024-10-01 00:23:18.647     St. Clair St & Erie St   
2024-09-30 23:42:11.207 2024-10-01 00:10:16.831  Ashland Ave & Chicago Ave   
2024-09-30 23:49:25.380 2024-10-01 00:06:27.476          900 W Harrison St   

                        start_station_id          end_station_name  \
started_at                                                           
2024-09-30 23:12:01.622           bdd4c3                       NaN   
2024-09-30 23:19:25.409              NaN    Benson Ave & Church St   
2024-09-30 23:32:24.672           9c619a  LaSalle St & Illinois St   
2024-09-30 23:42:11.207           72a04d    Loomis St & Archer Ave   
2024-09-30 23:49:25.380           11da85         900 W Harrison St   

                        end_station_id  start_lat  start_lng    end_lat  \
started_at                                                                
2024-09-30 23:12:01.622            NaN  42.012342 -87.688243  41.970000   
2024-09-30 23:19:25.409         a10cf0  42.070000 -87.730000  42.048214   
2024-09-30 23:32:24.672         fbd1ad  41.894345 -87.622798  41.890762   
2024-09-30 23:42:11.207         896337  41.895954 -87.667728  41.841633   
2024-09-30 23:49:25.380         11da85  41.874754 -87.649807  41.874754   

                           end_lng member_casual  trip_duration_min  
started_at                                                           
2024-09-30 23:12:01.622 -87.650000        casual          67.984200  
2024-09-30 23:19:25.409 -87.683485        casual          82.742067  
2024-09-30 23:32:24.672 -87.631697        member          50.899583  
2024-09-30 23:42:11.207 -87.657435        casual          28.093733  
2024-09-30 23:49:25.380 -87.649807        member          17.034933

#### Interpretation

Reflect on problem 2 and answer (briefly / concisely) the following questions:

1. What types did Pandas assign to `started_at` and `member_casual` in Task 2a? Why might these defaults be problematic?
2. Look at the values in the station ID fields. Based on what you learned about git commit IDs in HW3a, how do you think the station IDs were derived?
3. Explain in your own words what method chaining is, what `df.isna().sum()` does and how it works.
4. Assume you found ~10% missing values in station columns but ~0% in coordinates. What might explain this? How might you handle the affected rows?

##### Your Answers

### Interpretation 

**Data types:**  
In Task 2a, Pandas reads `started_at` and `member_casual` as `object`.
One problem of these defaults is that date columns now become strings and cannot be used to calculate elapsed time, while categorical data types use more memory and have slower comparisons.

**Station ID pattern:**  
Station IDs are short alphanumeric hashes similar to Git commit IDs.
These are likely created by hashing the station name or coordinates with the SHA or MD5 hash functions to create consistent keys without directly using the text.

**Method chaining & isna().sum():**  
In method chaining, one operation results from passing as an argument into the next operation inside the chain.
`df.isna().sum()` makes a DataFrame containing booleans showing where a value is missing. These booleans are totaled by column.

**Missing-value reasoning:**  
Station fields missing ten percent and coordinates missing zero percent, this may show that some trips began or ended with no station ID or name.
Solutions can include treating rows as "free-floating" trips or imputing or flagging the rows, depending on the analysis objective.

#### Follow-Up (Graduate Students Only)

Compare memory usage results in 2a.3 and 2b.3. What caused the change? Why are these numbers different from what is reported at the bottom of `df.info()`? Which should you use if data size is a concern?

Working with DataFrames typically requires 5-10x the dataset size in available RAM. On a system with 16GB, assuming about 30% overhead from the operating system and other programs, what range of dataset sizes would be safely manageable? Calculate using both 5x (optimistic) and 10x (conservative) multipliers, then explain which you'd recommend for reliable work.

##### Your Answers

### Follow-Up (Graduate Students Only)

**1) Memory usage: 2a.3 vs 2b.3**  
- **Reason for change:** Converting to a more efficient type reduces memory usage, e.g. `category` compresses repeated use of strings, while `datetime64[ns]` is more compact.
- **Why numbers differ from df.info():** The summary at the bottom of `df.info()` is an estimate and doesn't include deep memory of string objects.  
- **Which metric to use:**`df.memory_usage(deep=True).sum()`, which is a full, accurate measure of memory consumption.
  
**2) Manageable dataset size (16GB RAM, ~30% OS overhead)**  
- **Available for Python:** 16 × 0.70 = **11.2 GB**  
- **Optimistic (5×):** 11.2 / 5 = **~2.24 GB dataset**  
- **Conservative (10×):** 11.2 / 10 = **~1.12 GB dataset**

**Recommendation:** For safe reproducible analysis, use the conservative limit of around 1.1 GB for memory intensive operations with side-effects that duplicate the data in memory like merges, groupby, joins.

### Problem 3: Filtering and Transformation

With clean data loaded, you can now filter and transform it to answer specific questions. This problem focuses on Pandas' powerful indexing and filtering capabilities, along with creating derived columns that enable deeper analysis.

You'll continue working with the `rides` DataFrame from Problem 2, which has `started_at` set as the index.

#### Task 3a: Boolean Indexing and Membership Testing

Use boolean indexing and the `isin()` method to answer these questions:

1. How many trips were taken by *members* using *electric bikes*? Use `&` to combine conditions.
2. What percentage of all trips does this represent?
3. How many trips started at any of these three stations: "Streeter Dr & Grand Ave", "DuSable Lake Shore Dr & Monroe St", or "Kingsbury St & Kinzie St"? Use `isin()`.

Note: Remember to use parentheses around each condition when combining with `&`.

##### Your Code

In [20]:
# Trips by members using electric bikes
member_electric = rides[(rides["member_casual"] == "member") & (rides["rideable_type"] == "electric_bike")]
num_member_electric = len(member_electric)
print("Trips by members using electric bikes:", num_member_electric)

# Percentage of all trips this represents
percentage_member_electric = (num_member_electric / len(rides)) * 100
print(f"Percentage of all trips: {percentage_member_electric:.2f}%")

# Trips starting at any of the three given stations
stations = ["Streeter Dr & Grand Ave", "DuSable Lake Shore Dr & Monroe St", "Kingsbury St & Kinzie St"]
num_station_starts = len(rides[rides["start_station_name"].isin(stations)])
print("Trips starting at any of the three stations:", num_station_starts)


Trips by members using electric bikes: 33121
Percentage of all trips: 33.12%
Trips starting at any of the three stations: 2702


#### Task 3b: Create Derived Columns from Datetime

Add two categorical columns to the rides DataFrame based on trip start time:

1. `is_weekend`: Boolean column that is True for Saturday/Sunday trips. Use .dt.dayofweek on the index (Monday=0, Sunday=6).
2. `time_of_day`: String categories based on start hour:
   - "Morning Rush" if hour is 7, 8, or 9
   - "Evening Rush" if hour is 16, 17, or 18
   - "Midday" for all other hours

For step 2, initialize the column to "Midday", then use .loc[mask, 'time_of_day'] with boolean masks to assign rush hour categories. Extract hour using .dt.hour on the index.

After creating both columns, use value_counts() on time_of_day to show the distribution.

##### Your Code

In [21]:
# is_weekend: True for Saturday (5) or Sunday (6)
rides["is_weekend"] = rides.index.dayofweek.isin([5, 6])

# time_of_day: categorize by start hour
rides["time_of_day"] = "Midday"
rides.loc[rides.index.hour.isin([7, 8, 9]), "time_of_day"] = "Morning Rush"
rides.loc[rides.index.hour.isin([16, 17, 18]), "time_of_day"] = "Evening Rush"

# Show distribution of time_of_day
rides["time_of_day"].value_counts()


time_of_day
Midday          55912
Evening Rush    28218
Morning Rush    15870
Name: count, dtype: int64

#### Task 3c: Complex Filtering with query()

Use the `query()` method to find trips that meet **all** of these criteria:
- Casual riders (not members)
- Weekend trips  
- Duration greater than 20 minutes
- Electric bikes

Report:
1. How many trips match these criteria?
2. What percentage of all trips do they represent?
3. What is the average duration of these trips?

Hint: Column names work directly in `query()` strings. Combine conditions with `and`.

##### Your Code

In [22]:
# Filter trips meeting all criteria
casual_weekend_long_electric = rides.query(
    "member_casual == 'casual' and is_weekend == True and trip_duration_min > 20 and rideable_type == 'electric_bike'"
)

# Number of matching trips
num_trips = len(casual_weekend_long_electric)
print("Trips matching criteria:", num_trips)

# Percentage of all trips
percentage_trips = (num_trips / len(rides)) * 100
print(f"Percentage of all trips: {percentage_trips:.2f}%")

# Average duration
avg_duration = casual_weekend_long_electric["trip_duration_min"].mean()
print(f"Average duration (minutes): {avg_duration:.2f}")


Trips matching criteria: 1501
Percentage of all trips: 1.50%
Average duration (minutes): 40.37


#### Task 3d: Explicit Indexing Practice

Practice using `loc[]` and `iloc[]` for different selection tasks:

1. Use `iloc[]` to select the first 10 trips, showing only `member_casual`, `rideable_type`, and `trip_duration_min` columns
2. Use `loc[]` to select trips from October 15-17 (use date strings `'2024-10-15':'2024-10-17'`), showing the same three columns
3. Count how many trips occurred during this date range

Note: When using `iloc[]`, remember it's position-based (0-indexed). When using `loc[]` with the datetime index, you can slice using date strings.

##### Your Code

In [38]:
# First 10 trips using iloc[]
rides.iloc[:10][["member_casual", "rideable_type", "trip_duration_min"]]

# Trips from October 15–17 using loc[]
rides.loc["2024-10-15":"2024-10-17", ["member_casual", "rideable_type", "trip_duration_min"]]

# Display the result (Additional)
display(rides.iloc[:10][["member_casual", "rideable_type", "trip_duration_min"]])
display(rides.loc["2024-10-15":"2024-10-17", ["member_casual", "rideable_type", "trip_duration_min"]])


# Count trips in that date range
num_trips_oct_range = len(rides.loc["2024-10-15":"2024-10-17"])
print("Trips from Oct 15–17:", num_trips_oct_range)



member_casual  rideable_type  trip_duration_min
started_at                                                             
2024-09-30 23:12:01.622        casual  electric_bike          67.984200
2024-09-30 23:19:25.409        casual  electric_bike          82.742067
2024-09-30 23:32:24.672        member   classic_bike          50.899583
2024-09-30 23:42:11.207        casual   classic_bike          28.093733
2024-09-30 23:49:25.380        member  electric_bike          17.034933
2024-09-30 23:49:40.016        member  electric_bike          13.009367
2024-10-01 00:00:53.414        member   classic_bike           2.598817
2024-10-01 00:05:44.954        member  electric_bike           0.013433
2024-10-01 00:06:12.035        member  electric_bike          10.472933
2024-10-01 00:10:03.646        member   classic_bike           7.825683

member_casual  rideable_type  trip_duration_min
started_at                                                             
2024-10-15 00:00:12.781        casual  electric_bike           2.804233
2024-10-15 00:01:20.517        member  electric_bike          11.330867
2024-10-15 00:05:24.811        member  electric_bike           1.868850
2024-10-15 00:05:52.984        member  electric_bike           2.705083
2024-10-15 00:06:18.819        member  electric_bike           1.600867
...                               ...            ...                ...
2024-10-17 23:45:48.739        member  electric_bike          14.703100
2024-10-17 23:47:35.040        member  electric_bike          13.049867
2024-10-17 23:55:34.112        member   classic_bike           3.795400
2024-10-17 23:56:14.464        member  electric_bike          11.937217
2024-10-17 23:59:38.103        casual   classic_bike           5.266433

[7235 rows x 3 columns]

Trips from Oct 15–17: 7235


#### Interpretation

Reflect on this problem and answer (briefly / concisely) the following questions:

1. `isin()` advantages: Compare using `isin(['A', 'B', 'C'])` versus `(col == 'A') | (col == 'B') | (col == 'C')`. Beyond readability, what practical advantage does `isin()` provide when filtering for many values (e.g., 20+ stations)?
2. Conditional assignment order: In Task 3b, why did we initialize all values to "Midday" before assigning rush hour categories? What would go wrong if you assigned categories in a different order, or didn't set a default?
3. `query()` vs boolean indexing: The `query()` method in Task 3c could have been written with boolean indexing instead. When would you choose `query()` over boolean indexing? When might boolean indexing be preferable despite being more verbose?

##### Your Answers

### Interpretation

1- `isin([...])` gives greater speed through vectorized set-membership, accepts any iterable like lists, arrays, or Series, and keeps code maintainable if values change or expand. Chaining `==` with `|` is slow for writing if you have 20+ values, bug-prone with operator precedence, harder for reading it's not clear which `==` is for what column, and less maintainable.

2- The usual initialization of `"Midday"` gives a safe default, with subsequent overwrites giving rush hours. If the initialization is omitted, some of the rows will be uninitialized or overwritten even if other assignments are made. When a later mask overwrites an earlier category, it also overwrites part of earlier masks.

3- Use `query()` for nice, readable SQL-like queries with many conditions when you reference columns in the form of valid Python identifiers. Use boolean indexing with short boolean masks for simple conditions and when referencing Python objects, methods, complex functions, or non-valid Python identifiers in column names.


#### Follow-Up (Graduate Students Only)

Pandas supports a variety of indexing paradigms, including bracket notation (`df['col']`), label-based indexing (`loc[]`), and position-based indexing (`iloc[]`). The lecture recommended using bracket notation only for columns, and loc/iloc for everything else. Explain the rationale: why is this approach better than using bracket notation for everything, even though `df[0:5]` technically works for row slicing?

##### Your Answers

### Follow-Up (Grad Student)

Reasons to prefer `loc[]` / `iloc[]` notation over bracket notation:
Bracket notation (`df[...]`) is ambiguous. It selects columns when given a label, but it slices rows when given integers. This can lead to subtle bugs, especially in DataFrames with integer-based indexes that are interpreted as positional by default.

For clarity, use `loc[]` to specify an intent by label and `iloc[]` by position:
- `loc[]` always uses index labels (dates, strings, etc.).
- `iloc[]` always uses integer positions.

This makes code more readable with avoiding the risk of confusion when integer index values clash in integer labels plus gives predictable selection behavior in datasets.
`df[0:5]` to slice rows lacks the clarity of `df.iloc[0:5]`, and it acts in another way if the index is not a plain RangeIndex.

### Problem 4: Temporal Analysis and Export

Time-based patterns are crucial for understanding bike share usage. In this problem, you'll analyze when trips occur, how usage differs between rider types, and export filtered results. You'll use the datetime index you set in Problem 2 and the derived columns from Problems 2-3.

#### Task 4a: Identify Temporal Patterns

Use the datetime index to extract temporal components and identify usage patterns:

1. Extract the *hour* from the index and use `value_counts()` to find the most popular hour for trips. Report the peak hour and how many trips occurred during that hour.
2. Extract the *day name* from the index and use `value_counts()` to find the busiest day of the week. Report the day and number of trips.
3. Sort the results from step 2 to show days in order from Monday to Sunday (not by trip count). Use `sort_index()`.

Hint: Use `.dt.hour` and `.dt.day_name()` on the datetime index.

##### Your Code

In [40]:
# Check before using .hour or .day_name()
print(type(rides.index))

# Most popular hour for trips
hour_counts = rides.index.hour.value_counts()
peak_hour = hour_counts.idxmax()
peak_hour_count = hour_counts.max()
print(f"Peak hour: {peak_hour}:00 with {peak_hour_count} trips")

# Busiest day of the week
day_counts = rides.index.day_name().value_counts()
busiest_day = day_counts.idxmax()
busiest_day_count = day_counts.max()
print(f"Busiest day: {busiest_day} with {busiest_day_count} trips")

# Sort days Monday–Sunday
day_counts_sorted = day_counts.reindex([
    "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"
])
day_counts_sorted



<class 'pandas.core.indexes.datetimes.DatetimeIndex'>
Peak hour: 17:00 with 10574 trips
Busiest day: Wednesday with 16513 trips


started_at
Monday       11531
Tuesday      14970
Wednesday    16513
Thursday     16080
Friday       13691
Saturday     14427
Sunday       12788
Name: count, dtype: int64

#### Task 4b: Compare Groups with groupby()

Use `groupby()` (introduced in 07a) to compare trip characteristics across different groups:

1. Calculate the average trip duration by rider type (`member_casual`). Which group takes longer trips on average?
2. Calculate the average trip duration by bike type (`rideable_type`). Which bike type has the longest average trip?
3. Count the number of trips by rider type using `groupby()` with `.size()`. Compare this with using `value_counts()` on the `member_casual` column - do they give the same result?

Note: Use single-key groupby only (one column at a time).

##### Your Code

In [42]:
# Average trip duration by rider type
avg_duration_by_rider = rides.groupby("member_casual", observed=True)["trip_duration_min"].mean()
print(avg_duration_by_rider)
print("\nLongest average (rider type):", avg_duration_by_rider.idxmax())

# Average trip duration by bike type
avg_duration_by_bike = rides.groupby("rideable_type", observed=True)["trip_duration_min"].mean()
print(avg_duration_by_bike)
print("\nLongest average (bike type):", avg_duration_by_bike.idxmax())

# Trip counts by rider type: groupby().size() vs value_counts()
counts_groupby = rides.groupby("member_casual", observed=True).size()
counts_vc = rides["member_casual"].value_counts()

print("groupby size:\n", counts_groupby)
print("\nvalue_counts:\n", counts_vc)
print("\nSame result?", counts_groupby.equals(counts_vc.sort_index()))



member_casual
casual    23.978046
member    11.984493
Name: trip_duration_min, dtype: float64

Longest average (rider type): casual
rideable_type
classic_bike     20.337410
electric_bike    12.033618
Name: trip_duration_min, dtype: float64

Longest average (bike type): classic_bike
groupby size:
 member_casual
casual    34686
member    65314
dtype: int64

value_counts:
 member_casual
member    65314
casual    34686
Name: count, dtype: int64

Same result? True


#### Task 4c: Filter, Sample, and Export

Create a filtered dataset for weekend electric bike trips and export it:

The provided code once again uses Path to create an `output` directory and constructs the full file path as `output/weekend_electric_trips.csv`. Use the `output_file` variable when calling `.to_csv()`.

1. Filter for trips where `is_weekend == True` and `rideable_type == 'electric_bike'`
2. Use `iloc[]` to select the first 1000 trips from this filtered dataset
3. Use `reset_index()` to convert the datetime index back to a column (so it's included in the export)
4. Export to CSV with filename `weekend_electric_trips.csv`, including only these columns: `started_at`, `ended_at`, `member_casual`, `trip_duration_min`, `time_of_day`
5. Use `index=False` to avoid writing the default numeric index to the file

After exporting, report how many total weekend electric bike trips existed before sampling to 1000.

##### Your Code

In [43]:
# do not modify this setup code
from pathlib import Path

output_dir = Path('output')
output_dir.mkdir(exist_ok=True)
output_file = output_dir / 'weekend_electric_trips.csv'

# Task 4c code here...
# use the variable `output_file` as the filename for step 4

# Filter weekend electric bike trips
weekend_electric = rides[(rides["is_weekend"] == True) & (rides["rideable_type"] == "electric_bike")]

# Count total before sampling (for report)
total_weekend_electric = len(weekend_electric)

# Take first 1000 with iloc[]
weekend_electric_1k = weekend_electric.iloc[:1000].reset_index()

# Export selected columns to CSV
cols = ["started_at", "ended_at", "member_casual", "trip_duration_min", "time_of_day"]
weekend_electric_1k[cols].to_csv(output_file, index=False)

print("Total weekend electric bike trips (before sampling):", total_weekend_electric)
print("Saved:", output_file)



Total weekend electric bike trips (before sampling): 13026
Saved: output\weekend_electric_trips.csv


#### Interpretation

Reflect on this problem and answer the following questions:

1. `groupby() conceptual model`: Explain in your own words what `groupby()` does. Use the phrase "split-apply-combine" in your explanation and describe what happens at each stage.
2. `value_counts()` vs `groupby()`: In Task 4b.3, you compared two approaches for counting trips by rider type. When would you use `value_counts()` versus `groupby().size()`? Is there a situation where only one of them would work?
3. Index management for export: In Task 4c, why did we use `reset_index()` before exporting? What would happen if you exported with the datetime index still in place and used `index=False`?

##### Your Answers

### Interpretation

**`groupby() conceptual model` (split–apply–combine):** 

- **Split:** seperate the DataFrame into subsets based on group keys (e.g., `member_casual`).  
- **Apply:** run an operation independently on each subset (mean, size, agg).  
- **Combine:** assemble the per group results back into a single aligned object (Series/DataFrame).


**`value_counts()` vs `groupby().size()`:** 

- **`value_counts()`**: quickest for frequency of values in a **single Series** (e.g., `rides["member_casual"].value_counts()`), returns counts sorted by frequency by default.  
- **`groupby().size()`**: works from the **DataFrame** context and generalizes to one or more keys; integrates naturally with additional operations (`agg`, `mean`, etc.).  
- **Situations:** `value_counts()` only works on a single Series; `groupby().size()` works when counts are needed by a key while staying in DataFrame workflows or when extending to multi-key grouping.


**Index management for export:**  
`reset_index()` converts the datetime index into a regular column (`started_at`) so it is **included in the CSV**.  
If export occurs with the datetime index still in place and `index=False` is used, the index is **dropped** from the file—losing the start time information.  


#### Follow-Up (Graduate Students Only)

Compare `CSV` and _pickle_ formats for data storage and retrieval.

Pickle is Python's built-in serialization format that saves Python objects exactly as they exist in memory, preserving all data types, structures, and metadata. Unlike CSV (which converts everything to text), pickle is binary (not human readable) and maintains the complete state of your DataFrame. Also, pickle files only work in Python, while CSV is universal. Read more in the [Pandas documentation](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.to_pickle.html).

The code below investigates an interesting pattern: Do riders take longer trips from scenic lakefront stations even during rush hours? This could indicate tourists or recreational riders using these popular locations for leisure trips during typical commute times. The analysis filters for trips over 15 minutes that started from lakefront stations during morning (7-9am) or evening (4-6pm) rush hours, sorted by duration to see the longest trips first.

Run the code below, then answer the interpretation questions:

In [44]:
import os

# the following lines were commented out since they were run in 4c
# from pathlib import Path
# output_dir = Path('output')

# Ensure rides.index is a DatetimeIndex before using .hour
print(type(rides.index))  # should be DatetimeIndex

csv_file = output_dir / 'lakefront_rush_trips.csv'
pickle_file = output_dir / 'lakefront_rush_trips.pkl'

lakefront_rush = (
    rides
      .loc[
          (rides.index.hour.isin([7, 8, 9, 16, 17, 18])) &
          (rides['start_station_name'].str.contains('Lake Shore|Lakefront', case=False, na=False)) &
          (rides['trip_duration_min'] > 15)
      ]
      .sort_values('trip_duration_min', ascending=False)
      .head(1000)
      .reset_index()
      [['started_at', 'ended_at', 'start_station_name', 'end_station_name',
        'member_casual', 'rideable_type', 'trip_duration_min']]
)

print(f"Found {len(lakefront_rush)} long rush-hour trips from lakefront stations")

# Export to both formats
lakefront_rush.to_csv(csv_file, index=False)
lakefront_rush.to_pickle(pickle_file)

# Compare sizes
import os
csv_size = os.path.getsize(csv_file) / 1024  # KB
pickle_size = os.path.getsize(pickle_file) / 1024
print(f"\nCSV file size: {csv_size:.2f} KB")
print(f"Pickle file size: {pickle_size:.2f} KB")
print(f"Size difference: {abs(csv_size - pickle_size):.2f} KB")

# Load-time comparison
print("\nLoad time comparison:")
print("CSV:")
%timeit pd.read_csv(csv_file)
print("\nPickle:")
%timeit pd.read_pickle(pickle_file)

# Type preservation check
csv_loaded = pd.read_csv(csv_file)
pickle_loaded = pd.read_pickle(pickle_file)
print("\nData types from CSV (without parse_dates):")
print(csv_loaded.dtypes)
print("\nData types from Pickle:")
print(pickle_loaded.dtypes)


<class 'pandas.core.indexes.datetimes.DatetimeIndex'>
Found 498 long rush-hour trips from lakefront stations

CSV file size: 70.12 KB
Pickle file size: 22.18 KB
Size difference: 47.94 KB

Load time comparison:
CSV:
4.63 ms ± 441 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)

Pickle:
1.15 ms ± 81.1 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

Data types from CSV (without parse_dates):
started_at             object
ended_at               object
start_station_name     object
end_station_name       object
member_casual          object
rideable_type          object
trip_duration_min     float64
dtype: object

Data types from Pickle:
started_at            datetime64[ns]
ended_at              datetime64[ns]
start_station_name            object
end_station_name              object
member_casual               category
rideable_type               category
trip_duration_min            float64
dtype: object


After running the code, answer these questions:

1. Method chaining: The analysis uses method chaining with a specific formatting pattern:

   ```python
   result = (df
       .method1()
       .method2()
       .method3()
   )
   ```

   This wraps the entire chain in parentheses, allowing each method to appear on its own line without backslashes. Discuss why this makes formatting more readable, how it makes debugging easier, how it relates to seeing changes in the code with git diff, and what downsides heavy chaining might have.
3. Data types: Compare the dtypes from CSV versus pickle. What types were preserved by pickle that were lost in CSV? Why is this preservation significant for subsequent analysis?
4. Trade-offs: Given your observations about size, speed, and type preservation, when would you choose pickle over CSV for your work? When would CSV still be the better choice despite pickle's advantages?


### Follow-Up (Grad Student)

1- Method chaining (readability, debugging, git diff, downsides):
Enclosing each method call in parentheses, and placing that on a separate line, improves readability, gives clearer diffs due to the obvious line of the chain being changed at a glance, and makes debugging easier. Deep chaining may cause difficulty when inspecting intermediate states. Debugging errors within them may be difficult too.

2- Data types (CSV vs Pickle):
From the output:
- CSV: all columns loaded as `object`, except trip_duration_min (float64)
- Pickle: preserved `datetime64[ns]` for dates and rider and bike type as `category`
Pickle preservation is also important to avoid re-parsing and converting types for time-based filtering or comparison, and for efficient categorical processing.

3- Trade-offs:
- Pickle, which is faster, takes less space and preserves all Python data types, solves many issues of Marshal.
- Cons: Pickle is only suitable for Python, not human-readable.
- CSV is universally readable and supported by Excel, R, and SQL databases.
- Disadvantages include larger file size, slower input/output, and loss of types if not specified.
Use Pickle for intermediate analysis in a Python pipeline, and CSV when sharing or publishing data.

4- What did change from the original code with this fix? In your opinion, why didn't the original code work and why does this one work as intended?

It checks for an empty start key.
The rush hour filtering is simplified by using the datetime index (rides.index.hour) in a single .loc[] statement for the hour range, and applying the station name regex and the duration threshold in one statement. The output is then written and .reset_index() is used to convert the datetime index to a column.

Why the original didn't work/why this one works:
In earlier versions, the `.hour` attribute would not work if the index type was not a `DatetimeIndex`.
This checks the index type, uses the correct way to access datetimes, and performs all filtering in one compact and reproducible chain without partial filtering or NaN station mismatches.

## Reflection

Address the following questions in a markdown cell:

1. NumPy vs Pandas
   - What was the biggest conceptual shift moving from NumPy arrays to Pandas DataFrames?
   - Which Pandas concept was most challenging: indexing (loc/iloc), missing data, datetime operations, or method chaining? How did you work through it?
2. Real Data Experience
   - How did working with real CSV data (with missing values, datetime strings, etc.) differ from hw2b's synthetic NumPy arrays?
   - Based on this assignment, what makes Pandas well-suited for data analysis compared to pure NumPy?
3. Learning & Application
   - Which new skill from this assignment will be most useful for your own data work?
   - On a scale of 1-10, how prepared do you feel to use Pandas for your own projects? What would increase that score?
4. Feedback
   - Time spent: ___ hours (breakdown optional)
   - Most helpful part of the assignment: ___
   - One specific improvement suggestion: ___

### Your Answers

### Reflection

**NumPy vs Pandas**  
The biggest conceptual shift for me was realizing that Pandas isn’t just about numbers—it’s about labeled data.  
In NumPy, everything is position-based, but in Pandas I had to think in terms of columns, indices, and labels.  
The most challenging concept was indexing with `loc` and `iloc`, especially understanding when each one applies.  
I got more comfortable by experimenting in small examples and checking the outputs step by step.

**Real Data Experience**  
Working with real CSV data felt very different from the clean NumPy arrays in HW2b.  
Real data had missing values, inconsistent text, and datetime columns that needed parsing.  
It showed me that data cleaning is a big part of analysis.  
Pandas handled these issues smoothly—it can detect types, handle missing data, and perform complex operations in just a few lines.

**Learning & Application**  
The most useful new skill for me was learning how to filter and transform data using `loc[]`, `query()`, and method chaining.  
I also liked seeing how datetime indexing made time-based analysis feel natural.  
I’d rate my confidence using Pandas around **8/10**.  
I feel comfortable with the basics, but I want more practice with `groupby()` and larger datasets to reach a full 10.

**Feedback**  
Time spent: ~5 hours (3 hours coding, 1 hour debugging, 1 hour review and documentation).  
Most helpful part: Seeing how to use datetime data and boolean filters in real datasets.  
Improvement suggestion: Add a few short practice exercises on method chaining before the main assignment—it would make the transition smoother.
